bin/functions

```diff
@@ -47,67 +50,60 @@ function check-jq() {
-        done
-        for i in "${CACHE_DIR}"/adopt-*.json
-        do
-            [[ -e "$i" ]] || break
-            jq '(.[].release_name) |= sub("jdk-";"adopt-openjdk-") | (.[].release_name) |= sub("^jdk";"adopt-openjdk-")' \
-               "${i}" > "${i}.temp"
-            mv "${i}.temp" "${i}"
-        done
-    fi
-}
-
-function retrieve-sapmachine() {
-    args=('-s' '-f')
-    if [[ -n "${GITHUB_API_TOKEN:-}" ]]; then
-        args+=('-H' "Authorization: token $GITHUB_API_TOKEN")
-    fi
-    local cache_file="${CACHE_DIR}/sapmachine.json"
-    local filter='[
-.[] | select(.prerelease == false and .draft == false)
-| {
-  release_name: .name,
-  binaries: [ .assets[]
-    | select(.content_type == "application/x-tar")
-    | select(.name | startswith("sapmachine-jdk"))
-    | select(.name | endswith("linux-x64_bin.tar.gz") or endswith("osx-x64_bin.tar.gz"))
-    | {
-      package: {
-        link: .browser_download_url,
-        checksum_link: (.browser_download_url | sub("tar\\.gz$"; "sha256.txt"))
-      },
-      os: (if .name | endswith("osx-x64_bin.tar.gz") then "mac" else "linux" end),
-      architecture: "x64",
-      jvm_impl: "hotspot",
-      heap_size: "normal"
-    }
-  ]}
-]'
-    # shellcheck disable=SC2046
-    if [[ ! -r "${cache_file}" ]] || [[  $(set -- $(${STAT}) && echo "${1}") -le $(( $(date +%s) - 3600)) ]]
-    then
-        curl "${args[@]}" -L 'https://api.github.com/repos/SAP/SapMachine/releases' -o "${cache_file}"
-        jq "${filter}" "${cache_file}" > "${cache_file}.temp"
-        mv "${cache_file}.temp" "${cache_file}"
-    fi
+  local -a urls
+  local min_java=8
+  local max_java=14
+  local adopt_openjdk_url="https://api.adoptopenjdk.net/v3/assets/feature_releases/"
+  local github_curl_opts=("${CURL_OPTS[@]}" '-H' "Authorization: token $GITHUB_API_TOKEN")
+
+  for i in $(seq "${min_java}" "${max_java}")
+  do
+    local adopt_openjdk_options="&image_type=jdk&page=0&page_size=100&project=jdk&sort_order=ASC&vendor=adoptopenjdk"
+    local url="${adopt_openjdk_url}{${i}}/ga?architecture=${ARCHITECTURE}&os=${OS}${adopt_openjdk_options}"
+    urls+=("${url}")
+  done
+  urls+=('https://api.github.com/repos/SAP/SapMachine/releases')
+
+  if (( ${#CACHE_FILES[@]} == 0 )) || (( $(stat "${STAT_OPTS[@]}" "${CACHE_FILES[0]}") <= $(date +%s) - 3600 )) ; then
+    for url in "${urls[@]}"
+    do
+      case "${url}" in
+        *"api.adoptopenjdk.net"*)
+          curl "${CURL_OPTS[@]}" -L "${url}" -w "%{filename_effective}\n" \
+               -o "${CACHE_DIR}/adopt-#1.json" > /dev/null 2>&1
+          [[ "${url}" =~ \{([0-9]*)\} ]]
+          jq '(.[].release_name) |= sub("jdk-";"adopt-openjdk-") | (.[].release_name) |= sub("^jdk";"adopt-openjdk-")' \
+             "${CACHE_DIR}"/adopt-"${BASH_REMATCH[1]}".json > "${CACHE_DIR}"/adopt-"${BASH_REMATCH[1]}".temp
+          mv "${CACHE_DIR}"/adopt-"${BASH_REMATCH[1]}".temp "${CACHE_DIR}"/adopt-"${BASH_REMATCH[1]}".json
+          ;;
+        *"api.github.com"*)
```

Comment:
Should we more specific matching here?

For example, GraalVM and SAP Machine using same github api endpoint

Reply:
As soon as a more specific match is needed, we will make this more specific